### DISCUSSION TITLE
LFX Mentorship Summer 2022 Challenge

### DISCUSSION THREAD

**Discussion by alabulei1** [WasmEdge/WasmEdge#1442] [Show and tell]
### Introduction

For those who want to apply for LFX Mentorship Summer #1429, #1430, #1431, and #1433. This is a pretest for the application.

Please fork this repo and complete the challenge in your repo. Once you are done, post your repo’s GitHub link, as well as a screenshot demonstrating the customized host functions, and the rust SDK for creating your test cases, as a comment under this discussion.

### Requirements

1. Fork this repo and try to add brand new host functions. You should define and implement your host functions. FYR, I have created a template in https://github.com/WasmEdge/WasmEdge/tree/challenge/proposal. You can find the details in the last four commits.
2. After you finish the first step, you will need to create a wasm file to call your host functions. If you have no ideas about how to make this, please fork and refer to this example: https://github.com/second-state/wasmedge_hostfunctionexample_interface.
3. Then, once you implement both 1. and 2. You can use `wasmedge foo.wasm` to run your application.
4. Post your repo's GitHub link including the forked WasmEdge and the wasm file or the rust interface, a screenshot of the execution result, and your comments under this discussion.

Good luck!

**Comment by MediosZ** [comment]
Here's [the link of the repository](https://github.com/MediosZ/WasmEdge/tree/lfx-async) and the screenshot follows.
![Screenshot from 2022-05-13 19-18-25](https://user-images.githubusercontent.com/24819143/168272780-06dcfa00-b538-493b-8768-2fc0fa01827d.png)

**Can we pass `string` or `char *` to wasm program from host? I tried some solutions but failed.**

**Reply by abhinandanudupa** [reply]
I think you are supposed to add your own functions after learning from the example code

**Reply by MediosZ** [reply]
Thanks! I will do that later!

**Comment by TieWay59** [comment]
Hi @alabulei1, I finished my challenge.

Fork branch link: https://github.com/TieWay59/WasmEdge/tree/lfx-challenge-tieway59

Screenshot:
![image](https://user-images.githubusercontent.com/40034603/169876251-5b74cc85-9552-4ea1-85d9-4377f64c4f79.png)


Screenshot-alt:

```console
> Executing task: cargo run --package tw59challenge --bin tw59challenge <

   Compiling tw59challenge v0.1.0 (/root/WasmEdge/examples/tw59challenge)
    Finished dev [unoptimized + debuginfo] target(s) in 0.89s
     Running `/root/WasmEdge/examples/tw59challenge/../../tw59build/tools/wasmedge/wasmedge target/wasm32-wasi/debug/tw59challenge.wasm`
add Peter, student number = 1
add Paul, student number = 2
add Marry, student number = 3
Class ID: 101
Class Name: WasmEdge
Student: Peter
Student: Paul
Student: Marry
[src/main.rs:74] binary_gcd(2, 2) = 2
[src/main.rs:75] binary_gcd(2, 4) = 2
[src/main.rs:76] binary_gcd(12, 16) = 4
[src/main.rs:77] binary_gcd(111111111, 111111112) = 1
```

Document: https://github.com/TieWay59/2022-LFX-Mentorship-Tracker/blob/master/weekly-blogs/week-00-challenge.md 

I write this doc to help me track my operations, and it turns out to be very informative. I think it's not easy for a first-comer to finish this challenge simply by following a "commit history" of an outdated branch. If this doc tutorial is helpful, I can organize it into a more formal one.

BTW: Many thanks to @MediosZ because it spends me much time dealing with the breaking change of `ImportObject` in the latest master branch, so I reviewed your commit history and found the solution.

**Reply by MediosZ** [reply]
I am glad it helps~

**Reply by HLRJ** [reply]
Thank you so much！

**Comment by JialuGong** [comment]
Hi, I've finished my task, and this is [the link of my repo](https://github.com/JialuGong/WasmEdge), and [this](https://github.com/JialuGong/WasmEdge/tree/pretest/my_host_function/hfexample) is the interface code of this host function. 
A screenshot of the excution result is shown below.
![image](https://user-images.githubusercontent.com/45031678/170181693-83092746-a2ae-4833-a91d-3db573065b90.png)

**Comment by fengyangsi** [comment]
Hi @alabulei1, I finished my challenge.

Fork branch [link](https://github.com/fengyangsi/WasmEdge)

Screenshot:
![Screenshot_20220525_125316](https://user-images.githubusercontent.com/90323422/170208930-c85155f2-9f40-432a-8533-92ffb19ea7af.png)

Text:
```
this job run after 0 second
this job run after 1 second
this job run after 2 second
this job run after 3 second
this job run after 4 second
```

**Comment by leowei1234567** [comment]
I finished the challenge. 
one function for add class level implemented.

Here is the repo link: [easmedge](https://github.com/leowei1234567/WasmEdge)

Screenshot:
![tp1](https://user-images.githubusercontent.com/49530002/170237034-aebbfb27-ce95-4b45-ad1f-cca3183508d7.png)

**Comment by vinsiny** [comment]
this is my challenge fork[ resipository link](https://github.com/Vinson-Ben/WasmEdge/tree/challenge/vinson), I implement a host function-- add_teacher_name，the excution result screenshot:
<img width="857" alt="image" src="https://user-images.githubusercontent.com/29425022/170491846-e7c77a29-35ee-426b-884e-cd0c8ec8d9dd.png">

**Comment by 451846939** [comment]
I finished the challenge. [the link of the repository](https://github.com/451846939/WasmEdge/tree/challenge/huashao) and the screenshot follows.
![image](https://user-images.githubusercontent.com/35210535/170507272-de7c7c97-f296-437e-a5f5-80100b67b65b.png)

Many thanks to @MediosZ  and @TieWay59  My macos Monterey c++ build environment is giving me a headache, even I read the documentation [Build for Mac](https://wasmedge.org/book/en/extend/build_on_mac.html), so I looked at your build method and environment

**Comment by 2019zhou** [comment]
Hi,
I have defined and implemented my host functions called fibonacci with two functions "fibonacci_get_num", "fibonacci_get_series" and hope these are enough for the challenge.

screenshot:
![image](https://user-images.githubusercontent.com/52394242/170954710-9bc39f59-3332-427c-a40c-2b7604708275.png)



my repo link: https://github.com/2019zhou/WasmEdge
Thanks for the clear and detailed instructions. If there is anything else, please let me know. 😆

**Comment by stdrc** [comment]
Hello, I have created [second-state/wasmedge-anna](https://github.com/second-state/wasmedge-anna) and [second-state/wasmedge-anna-rs](https://github.com/second-state/wasmedge-anna-rs) early this year. These two repos contains a wrapper runtime with custom host functions and demo WASM applications that call the custom functions, respectively, which already covers the requirements of this pretest.
